### ISSUE TITLE
Feature request: preserve hard links

### ISSUE THREAD

**Issue by tylla** [syncthing/syncthing#3897]
I would like to request a feature to enable the special handling of hard links, namely overwrite them in-place.
I am aware that currently Syncthing doesn't care about hard links, it treats them as regular files and when updating such files it writes to a temp file, deletes the old file and renames the temp file. As a consequence every file that has multiple hard links gets detached from the original content and other links.
But sometimes there is a need to use content with multiple names (in separate directories), and this over-weighs the possible risk of file corruption that could rise from direct updating the content of such files.
I would suggest an option (e.g: `inPlaceOverwriteHardLinks` or something alike) that can be enabled on any share, so that files with link count greater than 1 are overwritten in place.

**Comment by AudriusButkevicius** [comment]
This makes very little sense when you take all the operating systems we support.

**Comment by calmh** [comment]
Yes. I'm not philosophically against doing this, but I suspect it would be complicated and fraught with corner cases given all the filesystems etc that don't support hard links. Complicated enough that I'm going to close this as it's extremely unlikely to be implemented.

**Comment by AudriusButkevicius** [comment]
I mean in place updates would solve this, but that already has a ticket somewhere.

**Comment by tylla** [comment]
@calmh I don't really see the corner cases. I mean the proposed logic would be:

- if there is no option enabled for the given share than do the usual way.
- if there is the option enabled, than check for multiple hard links
  - if check unsuccessful or single hard link found than do the usual way.
  - if the check indicates multiple hard links, than in-place update

If an operating system doesn't support hard links than the error handling would guide the flow of events into the right path.
The biggest part of this would be the implementation of in-place-update as this would be sticking out from the existing functionality and inflict some problems, but everyone using this feature would know that there is a trade-off between secure file update and multiple hard link usage.

@AudriusButkevicius could you point out the "in place update" ticket, I couldn't find anything like that. As a fallback that could do for me.

I would like to clarify the background of the request: there is a big directory of images, where one directory contains the images with original file names (IMG_XXXXX) and some other directories that contain sorted files with the same data but with more manageable names (like full date/time in the file name). As there are a few thousand images and directories nearing one hundred and constantly growing there isn't any other way to solve the storage of these files. Symlinks are out of question (as the main point is to keep the original if it gets deleted from the sorted directory) and bind mount doesn't fit the case.

**Comment by Ferroin** [comment]
@tylla The other problem though is that even if the OS supports hard-links, the FS might not.  The classic example is FAT{16,32}, which does not support hard-links but is supported on every platform Syncthing runs on.  There are a handful of other odd cases as well (classic HFS doesn't support hard-links either), and how the OS responds to trying to create a hard-link on a FS that doesn't support it isn't exactly well defined.

**Comment by AudriusButkevicius** [comment]
I can't find the ticket so opening this one instead.

**Comment by tylla** [comment]
@Ferroin I don't think there would be a problem if any OS/FS layer will not support the handling of hard links. In that case the feature will simply not work. But anyway if one user would like to use the "handle files with multiple hard links specially" feature then s/he will have to intentionally enable this, and I think we can assume that s/he will know which FS is used and whether that and the OS supports hard links.
Not to mention there isn't any place in the feature where Syncthing would/should create hard links at all, the whole point is to detect when there is hard linked content and treat it specially.
@AudriusButkevicius thanks for reopening/renaming. Meanwhile I have found the wiki article about [in-place-update](https://github.com/syncthing/syncthing/wiki/In-Place-Updates), maybe this is what you remembered.

**Comment by AudriusButkevicius** [comment]
If you feel like writing go, I think we could just have an option flag that instead of this:
https://github.com/syncthing/syncthing/blob/master/lib/model/rwfolder.go#L1464

Copies the content from file to file and then removes the temp file.

**Comment by calmh** [comment]
See also:

https://github.com/syncthing/syncthing/wiki/In-Place-Updates

**Comment by spectras** [comment]
> * if there is the option enabled, than check for multiple hard links
>     * if check unsuccessful or single hard link found than do the usual way.
>     * if the check indicates multiple hard links, than in-place update

This approach has a race condition built-in. It's totally possible that the second name for the file gets created after the check, but before the temp file is complete. What then? Calmh's link has a much better approach. Still has a race condition, but on a much shorted time span.

Yet the inherent danger of corrupting the files anytime makes in-place updates useless for most purposes.

**Comment by tylla** [comment]
Sorry for being away for so long, somehow I missed to "Watch" the issue.
@spectras Yeah, I am aware that it is not a perfect solution. And indeed the approach linked by @calmh is much better thought out.
What I tried to emphasize beforehand that this option is somewhat special in its nature, so I would say it could go with some warnings attached to it in the manual. (like don't use it to sync files that are concurrently modified on several sides)
To show some example: my use case is about me modifying my pictures at home in the evening and then being able to modify the same sync'ed pictures at work if I have some time. The hard link creation is like: a homebrew bash script copies the pictures from my SD card to an ".originals" directory and creates the hard links (with some extra tagging and proper naming like Date-Time.jpg) to them in a different "incoming" directory. The tagged pictures are moved by me or my wife to some other named directories, the bad ones are deleted, some are tagged further. These operations all keep the hard linking intact.
As I happen to have only computers and file systems where hard linking is available I could really use this feature for the above functionality, with whatever restrictions may apply.
@AudriusButkevicius I don't know if I can spare such time to try it, but anyway the file you linked is gone. What was it?
Is there currently an ongoing effort to implement the mechanism described in the Wiki?

**Comment by catb0t** [comment]
I figured I'd ask around, just in case something already exists or is possible for this purpose: [Atomically write a file without changing inodes (preserve hard link)
](https://unix.stackexchange.com/questions/478314/atomically-write-a-file-without-changing-inodes-preserve-hard-link)

I suspect the answer to that is "No, you must choose between atomicity and keeping inode numbers"

**Comment by 01010101lzy** [comment]
Just being curious - is there any update on this feature?

**Comment by AudriusButkevicius** [comment]
There hasn't been any comments for 2 years, hence I think it's easy to draw the conclusion from that.

**Comment by 01010101lzy** [comment]
Thanks, I’m just making sure that if I want to try implementing this feature someday, there won’t be another conflicting implementation being worked on already.


发件人: Audrius Butkevicius<mailto:notifications@github.com>
发送时间: 2021年1月1日 19:08
收件人: syncthing/syncthing<mailto:syncthing@noreply.github.com>
抄送: Rynco Maekawa<mailto:lynzrand@outlook.com>; Manual<mailto:manual@noreply.github.com>
主题: Re: [syncthing/syncthing] In place file updates (#3897)


There hasn't been any comments for 2 years, hence I think it's easy to draw the conclusion from that.

―

**Comment by AudriusButkevicius** [comment]
There isn't, but the implementation is not trivial, as it would require write ahead logs and things like that to do it safely. There was a wiki page somewhere discussing how this could be done.

**Comment by haoadoresorange** [comment]
Hi, today I ran to this issue. Mainly that I have a folder with different config files synced between devices, and a script on setup that put the hard links to those config files in place.
Update the config file in one machine doesn't update the config in others since the files are not updated in place, result in the link detached from the new content.

I have 2 questions regarding this feature:

1/ Why was the file not updated in place ? Just a design choice or there was technical reason ?
2/ In the first proposal, OP mentioned handling the hard link count, is it viable that in case the "option" is turned on, it just update everything in place regardless of  hard link problem.

**Comment by haoadoresorange** [comment]
A current work around is putting a watcher on inotify signals for that specific folder, and run the script on changes. But this requires adding the watcher on all machine.

**Comment by AudriusButkevicius** [comment]
You cannot update a file in place atomically. If the application is restarted half way through a write, you would end up with corrupted files, which has half new data, half old, and syncthing would have no idea if that is the true state the file should be in after restart (as for example user modified the file while syncthing was not running). Writing to a new file and moving/renaming it to place is atomic.

**Comment by haoadoresorange** [comment]
Is atomicity abosulutely necessary in every case ? Afaik other solutions (such as Dropbox) write files in place. I understand that it is crucial in some cases, but most of the time I only sync text configuration files.
Do you think having an option is viable ?

**Comment by AudriusButkevicius** [comment]
This requires quite a lot of moving parts to do sensibly. 

You cannot know if the file is in some partial state because the user modified the file, or if because the process was terminated. The fact that you marked that the file needs syncing doesn't help in the case where the user actually modifies the file while the application is not running, because you'd just discard/overwrote/revert the users changes after restart, causing data loss.

So you need atomicity to avoid corruption.

Just having "an option because I sync text files" is not an option. If this was solved, it would have to be solved sensibly.

**Comment by haoadoresorange** [comment]
I see, was there any viable solution to this ? Or is there an API for hooks-alike ?

**Comment by calmh** [comment]
There is an API. I'm not sure what "this" you're looking for a solution to?

**Comment by haoadoresorange** [comment]
Sorry, I was referring to the detachment from new content of hard links that got linked to the files in a synced folder. (because a tempo file replaces the original so its not the same inode anymore)

**Comment by calmh** [comment]
Yeah as noted that’s how Syncthing works, so there’s no workaround for that.

**Comment by spectras** [comment]
I would just like to add that the way of dealing with files that syncthing uses is standard on Unix systems. Pretty much all modern applications work that way (writing to a new file then atomically renaming it over the old one).

It is the only way to prevent losing user data in case of a crash or a power outage: assuming proper filesystem settings, at any point in time one can recover either the old file or the new one (or both). On the other hand, modifying file in place can leave you with an empty file or partial file and your data irremediably lost.

That's okay if you have another copy to recover. But for many syncthing users, syncthing **is** the tool that maintains such copies. Dependability is key: if it suddenly started to lose files, most of its value would go away.

**Comment by tylla** [comment]
@spectras I think there is no debate about the defaultness (and usefulness) of write-new-file-and-rename-over-old method.
The whole point here is about a possible feature covering those rare but painful to handle cases where one has to use hard-links (for whatever reasons) and sync them between multiple machines.
It would be the typical thing that comes with strings attached, so off by default, but if needed then one could use it while knowing the possible risks associated.

**Comment by calmh** [comment]
The feature request does not require arguing for, it's clear. It requires implementation. There's an embryo of a design early in this thread. It just lacks someone willing to tackle it.

**Comment by haoadoresorange** [comment]
Maybe changing the name of the issuet to something like, Detached hard link due to in place file updates ?

**Comment by AudriusButkevicius** [comment]
I disagree. I think you are describing only one of the symptoms.

**Comment by haoadoresorange** [comment]
just a quick note on this, finally for me making the scripts deal with symlink to the files was a more trivial task than proposing in place write for syncthing

**Comment by EternityForest** [comment]
One very useful application for this would be databases. If every node could have its own log file where it stored operations, you could implement a CRDT like cr-sqlite, in a way that allows sync.

You could also manage text based log files very effectively, and it would be useful for syncing video files in real time as they are recorded.

This wouldn't actually require full in place updates, just in place file append operations, with fallback to normal sync, and with the understanding that only one device is going to be appending to the file. 

For this limited use case, perhaps an options dotfile could go right in the same folder with the log files, containing a list of file patterns to be treated specially.

Applications could then create that file automatically, if they were SyncThing-aware, or you could manually add it.

**Comment by pizdjuk** [comment]
Right now I ran into the issue that there is no simple option to have some folder only to share it with a devices, where there is not much space.

For example, there is some big folder to share. And in that folder there are only some files that I want to see on my android device. I would simply create a subfolder and populate it with hardlinks to files I want. But that if I change some file from the android the hardlink is unlinked. Clear, I could set the folder as "read only", but anyway... would be a useful feature.
